core/src/avm2/globals/flash/ui/context_menu.rs

```diff
@@ -1,19 +1,127 @@
+            items.set_property(
+                &Multiname::public("forwardAndBack"),
+                Value::Bool(false),
+                activation,
+            )?;
+            items.set_property(&Multiname::public("loop"), Value::Bool(false), activation)?;
+            items.set_property(&Multiname::public("play"), Value::Bool(false), activation)?;
+            items.set_property(&Multiname::public("print"), Value::Bool(false), activation)?;
+            items.set_property(
+                &Multiname::public("quality"),
+                Value::Bool(false),
+                activation,
+            )?;
+            items.set_property(&Multiname::public("rewind"), Value::Bool(false), activation)?;
+            items.set_property(&Multiname::public("save"), Value::Bool(false), activation)?;
+            items.set_property(&Multiname::public("zoom"), Value::Bool(false), activation)?;
+        }
+    }
 
     Ok(Value::Undefined)
 }
+
+pub fn make_context_menu_state<'gc>(
+    menu: Option<Object<'gc>>,
+    activation: &mut Activation<'_, 'gc, '_>,
+) -> context_menu::ContextMenuState<'gc> {
+    let mut result = context_menu::ContextMenuState::new();
+
+    macro_rules! check_bool {
+        ( $obj:expr, $name:expr, $value:expr ) => {
+            matches!(
+                $obj.get_property(&Multiname::public($name), activation),
+                Ok(Value::Bool($value))
+            )
+        };
+    }
+
+    let mut builtin_items = context_menu::BuiltInItemFlags::for_stage(activation.context.stage);
+    if let Some(menu) = menu {
+        if let Ok(Value::Object(builtins)) =
+            menu.get_property(&Multiname::public("builtInItems"), activation)
+        {
+            if check_bool!(builtins, "zoom", false) {
+                builtin_items.zoom = false;
+            }
+            if check_bool!(builtins, "quality", false) {
+                builtin_items.quality = false;
+            }
+            if check_bool!(builtins, "play", false) {
+                builtin_items.play = false;
+            }
+            if check_bool!(builtins, "loop", false) {
+                builtin_items.loop_ = false;
+            }
+            if check_bool!(builtins, "rewind", false) {
+                builtin_items.rewind = false;
+            }
+            if check_bool!(builtins, "forwardAndBack", false) {
+                builtin_items.forward_and_back = false;
+            }
+            if check_bool!(builtins, "print", false) {
+                builtin_items.print = false;
+            }
+        }
+    }
+
+    result.build_builtin_items(builtin_items, activation.context.stage);
+
+    if let Some(menu) = menu {
+        if let Ok(Value::Object(custom_items)) =
+            menu.get_property(&Multiname::public("customItems"), activation)
+        {
+            // note: this borrows the array, but it shouldn't be possible for
+            // AS to get invoked here and cause BorrowMutError
```

Comment:
The type checks on this property is just that it's an Array; the contents are unbounded. Presumably it should be possible to chuck a custom class in there with a `caption` property that tries to do naughty things.

However, this is really, *really* unlikely to actually be a problem.

Reply:
FP still does a typecheck (or rather, I'd guess they do a native-object-check?) so this can't happen on normal SWFs. Do you want me to do a ContextMenuItem check here too?

Reply:
That might be a good idea.